src/Oscoin/Node/Mempool/Class.hs

```diff
@@ -8,13 +8,15 @@ module Oscoin.Node.Mempool.Class
 import           Oscoin.Prelude
 
 import           Oscoin.Crypto.Hash (Hashed)
+import           Oscoin.Data.Tx (TxValidationError)
 import           Oscoin.Node.Mempool.Event (Channel, Event(..))
 
 -- MonadMempool ---------------------------------------------------------------
 
 class Monad m => MonadMempool c tx m | m -> tx, m -> c where
-    -- | Add transactions to the mempool, notifying all subscribers.
-    addTxs :: Foldable t => t tx -> m ()
+    -- | Add a transaction to the mempool if it is valid.
+    -- Notifies all subscribers.
+    addTx :: tx -> m (Either (TxValidationError c tx) ())
```

Comment:
But this would prevent us from defining generic instances of `Mempool` that work with any `tx`. For example we would need to provide two instances for `Tx` and `OscoinTx` which would have the exact same method definitions.

Reply:
Oh, right, sorry, still half awake. Yeah, the fundeps here are not really doing us a favour, as they are locking in place the `tx` given the concrete monad 😏